Feature/control scheme

rkbhkp opened the pull request:
This pull request of the feature/control-scheme should include everything needed to connect controllers, select schemes,  and drive the rover/send packets

rkbhkp committed:
Made the file

Made the file deleted some stuff from the gps which I copied for it.

rkbhkp committed:
added the two buttons to the controlscheme

added the buttons and the text for the controlscheme, still needs to be formated correctly, drop down bars need added, and apility to actually switch what is being done

rkbhkp committed:
Changed what is displayed on the buttons

rkbhkp committed:
Fixed the toggle of the buttons

the buttons for the controler scheme now toggle on and off, but there is a green highlight

rkbhkp committed:
Temporary Submition, working on adding controller functionality

Fixed the green color added drop down and on/off buttons still needs controller functionality and formating

rkbhkp committed:
Merge branch 'dev' into feature/Control-Scheme

rkbhkp committed:
Revert "Temporary Submition, working on adding controller functionality"

This reverts commit 9a605490afdc89d6d2c64187009df72980283f53.

rkbhkp committed:
fixed the layout for the control scheme

still needs controller support added

rkbhkp committed:
Merge branch 'dev' into feature/Control-Scheme

rkbhkp committed:
fixed conflicts

rkbhkp committed:
Added controller functionality

Controllers will now connect and accept input

rkbhkp committed:
Added controllerinput schemes and a differnet approach to handling controllers

added our own version of controller supprot that uses a 100 ms clock rather than using tick(). Also added controllerinput.tsx that will allow us to have different controller schemes

rkbhkp committed:
Made it so the controllers activate once the on button is cliked

Changed it so the controller inputs will start being tracked once the on button is clikced in the control scheme

rkbhkp committed:
Changes made when testing, sends commands to the rover

Changes made when doing testing. The rover can now accept the inputs from the controller. Also started work on other control schemes

rkbhkp committed:
Update package.json

rkbhkp committed:
Added a list of what each button does with the controller

rkbhkp committed:
MVP control Scheme

thomas-gleiforst committed:
Merge dev into control scheme

rkbhkp committed:
Changed what the selections display, flight stick implemtation

added the controls for the flight stick and made it so that only the controls that go with the certain type of controller display when selected

rkbhkp committed:
Removed gamepaditem file

thomas-gleiforst committed:
Touched up control scheme and add drive

Drive does need the negative drive to look like it is going backwards. There is a style there that is "supposed" to do it, but it doesn't

thomas-gleiforst committed:
Gimbal view

rkbhkp committed:
Fixed naming stuff for flightstick

rkbhkp committed:
Fixed Scheme problem

it will now auto select a valid theme if none will set the scheme to "" so that it doesnt give errors,

thomas-gleiforst committed:
Merge branch 'dev' into feature/Control-Scheme

thomas-gleiforst committed:
Merge dev into control scheme

thomas-gleiforst committed:
Thoroughly update drive comments

Waiting on people to agree that r can be simplified in this way, but the math seems to check out. Also simplied the fraction for  (-4 / pi) * (theta + (5 / 4) * pi) + 2 to something that makes more sense and follows a pattern

thomas-gleiforst commented on Electron/src/Core/components/ControlScheme.tsx:
```diff
@@ -0,0 +1,291 @@
+// eslint-disable-next-line max-classes-per-file
+import React, { Component, ReactNode, useState } from "react"
```
Remove unused imports (ReactNode, useState)

thomas-gleiforst commented on Electron/src/Core/components/ControlScheme.tsx:
```diff
@@ -0,0 +1,291 @@
+// eslint-disable-next-line import/no-mutable-exports
+export let inputs: any = {}
+
+function controller(passedScheme: any, pos: any): any {
```
Leave a comment describing what the passed in scheme is and what this function aims to accomplish

thomas-gleiforst commented on Electron/src/Core/components/ControlScheme.tsx:
```diff
@@ -0,0 +1,291 @@
+      }
+    }
+
+    const FlightStickIndex = controllerList.findIndex(element => {
+      if (element && element.indexOf("Logitech Extreme 3D") >= 0) return true
+      else return false
+    })
```
This findIndex is a pretty complicated, important, and precise function - be sure to comment heavily

thomas-gleiforst commented on Electron/src/Core/components/ControlScheme.tsx:
```diff
@@ -0,0 +1,291 @@
+            inputs[button] = 0.0
+          } else {
+            inputs[button] *= -1
+          }
```
Comment to explain the difference between buttons and thumbsticks, that we set to zero because of the deadzon, and multiply by -1 because thumbsticks are all inverted from the way you normally think of them

thomas-gleiforst commented on Electron/src/Core/components/ControlScheme.tsx:
```diff
@@ -0,0 +1,291 @@
+          }
+        }
+      }
+      console.log(inputs)
```
Remove before merge into dev

thomas-gleiforst commented on Electron/src/Core/components/ControlScheme.tsx:
```diff
@@ -0,0 +1,291 @@
+class ControlScheme extends Component<IProps, IState> {
+  constructor(props: Readonly<IProps>) {
+    super(props)
+    this.state = {
+      functionality: {
+        Drive: {
+          toggled: "Off",
+          scheme: "TankDrive",
+          controller: "Xbox 1",
+          interval: null,
+        },
+        MainGimbal: {
+          toggled: "Off",
+          scheme: "Gimbal",
+          controller: "Xbox 2",
+          interval: null,
+        },
+      },
```
Think about either expanding this to have defaults for science and arm so we can use one control scheme across the whole code base, or take these in from IProps

rkbhkp replied:
added arm and science

thomas-gleiforst commented on Electron/src/Core/components/ControlScheme.tsx:
```diff
@@ -0,0 +1,291 @@
+        break
+      }
+    }
+    console.log(defaultScheme)
```
Remove console.logs before merge into dev

thomas-gleiforst commented on Electron/src/Core/components/ControlScheme.tsx:
```diff
@@ -0,0 +1,291 @@
+
+  controllerChange(event: { target: { value: string } }, config: string): void {
+    let defaultScheme = ""
+    for (const scheme in CONTROLLERINPUT) {
+      if (
+        CONTROLLERINPUT[scheme].config === config &&
+        event.target.value.indexOf(CONTROLLERINPUT[scheme].controller) >= 0
+      ) {
+        defaultScheme = scheme
+        break
+      }
+    }
```
Better comment / explain why we are doing this - we want to default to a useful scheme so that a controller's shown value in the selector is actually "selected" and being used (if there is one)

thomas-gleiforst commented on Electron/src/Core/components/ControlScheme.tsx:
```diff
@@ -0,0 +1,291 @@
+          },
+        },
+      },
+      () => {
+        if (this.state.functionality[config].toggled === "On") {
+          this.setState({
+            functionality: {
+              ...this.state.functionality,
+              [config]: {
+                ...this.state.functionality[config],
+                interval: controller(
+                  this.state.functionality[config].scheme,
+                  this.state.functionality[config].controller
+                ),
+              },
+            },
+          })
+        }
```
Potentially explain why we are using this call back for this.setState to properly handle clearing and setting a new interval

thomas-gleiforst commented on Electron/src/Core/components/ControlScheme.tsx:
```diff
@@ -0,0 +1,291 @@
+  }
+
+  buttonToggle(config: string): void {
+    inputs = {}
```
See if you need to add this into scheme change and controller change as well

rkbhkp replied:
added it to scheme and controller functions

thomas-gleiforst commented on Electron/src/Core/components/ControlScheme.tsx:
```diff
@@ -0,0 +1,291 @@
+      <div style={this.props.style}>
+        <div style={label}>Control Scheme</div>
+        <div style={container}>
+          {/* [{name: Drive}].map({return(this.state.controllerInput[name].controller = value)})["Xbox 1", "Xbox 2", "Xbox 3", "Flight Stick"] */}
```
Remove this kind of comment before merging into dev

rkbhkp replied:
removed

thomas-gleiforst commented on Electron/src/Core/components/ControlScheme.tsx:
```diff
@@ -0,0 +1,291 @@
+        <div style={label}>Control Scheme</div>
+        <div style={container}>
+          {/* [{name: Drive}].map({return(this.state.controllerInput[name].controller = value)})["Xbox 1", "Xbox 2", "Xbox 3", "Flight Stick"] */}
+          {["Drive", "MainGimbal"].map(config => {
```
Take this in from props

thomas-gleiforst commented on Electron/src/Core/components/ControlScheme.tsx:
```diff
@@ -0,0 +1,291 @@
+      }
+      console.log(inputs)
+    }
+  }, 100)
```
Maybe decrease this to like 50ms

rkbhkp replied:
changed to 50

thomas-gleiforst commented on Electron/src/Core/components/ControlScheme.tsx:
```diff
@@ -0,0 +1,291 @@
+    }
+  }
+
+  render(): JSX.Element {
```
The styling may need to be touched up a little more - things are a little uneven, like the selectors are more than willing to not line up

thomas-gleiforst commented on Electron/src/Core/components/ControllerInput.tsx:
```diff
@@ -0,0 +1,108 @@
+    3: right stick up/down
+*/
+
+export const CONTROLLERINPUT = {
```
This could very easily be made into a json

thomas-gleiforst commented on Electron/src/Core/components/ControllerInput.tsx:
```diff
@@ -0,0 +1,108 @@
+      VectorY: { buttonType: "joystick", buttonIndex: 1 },
+      Throttle: { buttonType: "joystick", buttonIndex: 6 },
+      ForwardBump: { buttonType: "button", buttonIndex: 8 },
+      BackwardBump: { buttonType: "button", buttonIndex: 9 },
```
I don't think these indexes are correct for the flight controller, or it handles buttons differently, because when tested on rover it continuously spams forward bump

rkbhkp replied:
Fixed, had a logic error so it was always sending drive commands. Also, some of the buttons may be sticking on the actual controller but the current ones work.

thomas-gleiforst commented on Electron/src/RED/components/Gimbal.tsx:
```diff
@@ -0,0 +1,95 @@
+const UpArrow = path.join(__dirname, "../assets/UpArrow.png")
+const DownArrow = path.join(__dirname, "../assets/DownArrow.png")
+
+class Gimbal extends Component<IProps, IState> {
```
This file should probably be commented

thomas-gleiforst commented on Electron/src/Core/components/ControlScheme.tsx:
```diff
@@ -0,0 +1,291 @@
+}
+
+// eslint-disable-next-line import/no-mutable-exports
+export let inputs: any = {}
```
name this something more descriptive like controllerInputs

rkbhkp replied:
Changed it to controllerInputs in ControlScheme.tsx

thomas-gleiforst commented on Electron/src/RED/components/Drive.tsx:
```diff
@@ -0,0 +1,170 @@
+      // We base our speed off the value of the larger vector component
+      const r = Math.max(Math.abs(x), Math.abs(y))
+      // Scale vector is explained in terms of the right wheels, and the left wheels can be thought of as
+      // a reflection over the y axis, which is easiest obtained by adjusting the angle 45deg and inverting the result
```
*adjusts angles 90deg

thomas-gleiforst commented on Electron/src/RED/components/Drive.tsx:
```diff
@@ -0,0 +1,170 @@
+            <progress
+              value={this.state.leftSpeed < 0 ? -this.state.leftSpeed : 0}
+              max={1000}
+              style={{ display: "block", float: "right" }}
```
Maybe try rotating 180deg

thomas-gleiforst commented on Electron/src/RED/components/Gimbal.tsx:
```diff
@@ -0,0 +1,95 @@
+      controlling = "Drive"
+      image = DownArrow
+    }
+    if ("PanLeft" in inputs && "TiltLeft" in inputs) {
```
Maybe if on PanRight and TiltRight aswell

thomas-gleiforst commented on Electron/src/RED/components/Gimbal.tsx:
```diff
@@ -0,0 +1,95 @@
+      image = DownArrow
+    }
+    if ("PanLeft" in inputs && "TiltLeft" in inputs) {
+      if (controlling === "Main") {
+        rovecomm.sendCommand("LeftMainGimbal", [inputs.PanLeft * -5, inputs.TiltLeft * 5])
+        rovecomm.sendCommand("RightMainGimbal", [inputs.PanRight * -5, inputs.TiltRight * 5])
+      } else if (controlling === "Drive") {
+        rovecomm.sendCommand("LeftDriveGimbal", [inputs.PanLeft * 5, inputs.TiltLeft * -5])
+        rovecomm.sendCommand("RightDriveGimbal", [inputs.PanRight * 5, inputs.TiltRight * -5])
```
fix these multiples to better match Valkyrie, also drive takes [tilt, discard]

thomas-gleiforst commented on Electron/src/RED/components/Gimbal.tsx:
```diff
@@ -0,0 +1,95 @@
+      <div style={this.props.style}>
+        <div style={label}>Gimbal</div>
+        <div style={container}>
+          <img src={this.state.image} alt={this.state.controlling} />
```
maybe constrict size?

thomas-gleiforst commented on Electron/src/Core/components/ControlScheme.tsx:
```diff
@@ -0,0 +1,291 @@
+// eslint-disable-next-line import/no-mutable-exports
+export let inputs: any = {}
+
+function controller(passedScheme: any, pos: any): any {
```
Look into if there is an event that triggers whenever the state of a gamepad has changed instead of needing this interval

rkbhkp replied:
added to issues, not resolved

thomas-gleiforst commented on Electron/src/RED/components/Drive.tsx:
```diff
@@ -0,0 +1,170 @@
+              style={{ display: "block", float: "right" }}
+            />
+            <div style={{ width: "30%", textAlign: "center" }}>Left Speed: {this.state.leftSpeed}</div>
+            <progress style={{ flexGrow: 1 }} value={this.state.leftSpeed > 0 ? this.state.leftSpeed : 0} max={1000} />
```
Maybe make this max a part of state

thomas-gleiforst commented on Electron/src/RED/components/Drive.tsx:
```diff
@@ -0,0 +1,170 @@
+            />
+          </div>
+          <div style={row}>
+            Speed Limit:
```
Ensure 0 <= speedLimit <= 1000

rkbhkp committed:
Changed the inputs name to controllerInput

rkbhkp committed:
Fixed the problem where the flight stick wouldnt give correct values

Fixed logic error that had it always had the bump sending drive commands, also fixed variable names in drive and gimbal

thomas-gleiforst committed:
Add comments to drive and gimbal, clean up styling, set bounds on speedlimit, rotate progress bars

thomas-gleiforst committed:
Tank drive left and right wheels were swapped

rkbhkp committed:
removed  console.logs

rkbhkp committed:
changed refreshrate of controller input to 50 instead of 100

rkbhkp committed:
nows clears the inputs when changing controllers or schemes, changed deadzone from .15 to .05

all that plus removed an old comment

rkbhkp committed:
Made it so that the configs are taken in as a list of strings  in props

rkbhkp committed:
added defaults for arm and science

rkbhkp committed:
changed the controlerInput into a json

rkbhkp committed:
added comments for the controller input

rkbhkp committed:
added comment for the default controller selection

rkbhkp committed:
added comment for the controller function

rkbhkp committed:
added comments for handling the flightstick

rkbhkp committed:
added comment for the set state and clearing interval

thomas-gleiforst committed:
Update control scheme to add science
